go-controller/pkg/ovn/egressip.go

```diff
@@ -0,0 +1,542 @@
+				// this watcher receives a delete.
+				klog.V(5).Infof("EgressIP: %s update for pod: %s in namespace: %s, needs retry: %v", eIP.Name, newPod.Name, namespace.Name, oc.modeEgressIP.needsRetry(newPod))
+				if oc.modeEgressIP.needsRetry(newPod) {
+					if err := oc.modeEgressIP.addPod(eIP, newPod); err != nil {
+						klog.Errorf("Unable to add pod: %s/%s to EgressIP: %s, err: %v", newPod.Namespace, newPod.Name, eIP.Name, err)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				pod := obj.(*kapi.Pod)
+				// FYI: we can be in a situation where we processed a pod ADD for which there was no IP
+				// address assigned. If the pod is deleted before the IP address is assigned,
+				// we should not process that delete (as nothing exists in OVN for it)
+				klog.V(5).Infof("EgressIP: %s has stopped matching on pod: %s in namespace: %s, needs delete: %v", eIP.Name, pod.Name, namespace.Name, !oc.modeEgressIP.needsRetry(pod))
+				if !oc.modeEgressIP.needsRetry(pod) {
+					if err := oc.modeEgressIP.deletePod(eIP, pod); err != nil {
+						klog.Errorf("Unable to delete pod: %s/%s to EgressIP: %s, err: %v", pod.Namespace, pod.Name, eIP.Name, err)
+					}
+				}
+			},
+		}, nil)
+	if err != nil {
+		return err
+	}
+	oc.egressIPPodHandlerCache[getEgressIPKey(eIP)] = *h
+	return nil
+}
+
+func (oc *Controller) deleteNamespaceEgressIP(eIP *egressipv1.EgressIP, namespace *kapi.Namespace) error {
+	oc.egressIPPodHandlerMutex.Lock()
+	defer oc.egressIPPodHandlerMutex.Unlock()
+	delete(oc.egressIPPodHandlerCache, getEgressIPKey(eIP))
+	if err := oc.deleteAllEgressPods(eIP, namespace); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (oc *Controller) deleteAllEgressPods(eIP *egressipv1.EgressIP, namespace *kapi.Namespace) error {
+	pods, err := oc.kube.GetPods(namespace.Name, eIP.Spec.PodSelector)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		if !oc.modeEgressIP.needsRetry(&pod) {
+			if err := oc.modeEgressIP.deletePod(eIP, &pod); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (oc *Controller) assignEgressIPs(eIP *egressipv1.EgressIP) ([]egressipv1.EgressIPStatus, error) {
+	oc.eIPAllocatorMutex.Lock()
+	defer oc.eIPAllocatorMutex.Unlock()
+
+	if len(oc.eIPAllocator) == 0 {
+		oc.egressAssignmentRetry.Store(eIP, true)
+		return nil, fmt.Errorf("no assignable nodes")
+	}
+
+	assignments := []egressipv1.EgressIPStatus{}
+	eNodes, existingAllocations := oc.getSortedEgressData()
+	klog.V(5).Infof("Current assignments are: %+v", existingAllocations)
+	for _, egressIP := range eIP.Spec.EgressIPs {
+		klog.V(5).Infof("Will attempt assignment for egress IP: %s", egressIP)
+		eIPC := net.ParseIP(egressIP)
+		if eIPC == nil {
+			klog.Errorf("Unable to parse provided EgressIP: %s, invalid", egressIP)
+			continue
+		}
+		for i := 0; i < len(eNodes); i++ {
+			klog.V(5).Infof("Attempting assignment on egress node: %+v", eNodes[i])
+			if _, exists := existingAllocations[eIPC.String()]; !exists {
```

Comment:
this is a lot of nested statements. Maybe consider breaking some up by inversing the logic and using continue